### PR TITLE
Fix run-checker ubsan build

### PR DESCRIPTION
**Pull request by mattcaswell**
The ubsan run-checker build is configured as follows:

````
CC=clang ../openssl/config -d --strict-warnings enable-ubsan -DPEDANTIC -DOPENSSL_SMALL_FOOTPRINT -fno-sanitize=alignment
````
This PR contains 2 commits to address 2 different issues with the above config line:

1) The first commit addresses ubsan failures. Addition using the NULL pointer (even when adding 0) is undefined
behaviour. Recent versions of ubsan are now complaining about this, so we fix various instances.

2) The second commit address an OPENSSL_SMALL_FOOTPRINT issue. If OPENSSL_SMALL_FOOTPRINT was defined then the CHACHA20-POLY1305 implementation for TLS went down a different codepath that failed to adjust the payload length to remove the tag.

Thanks to @kroeckx for reporting the PEM ubsan issue.

### CODE REVIEW

File: crypto/bio/bss_mem.c [hunk 1]
```diff
@@ -299,7 +299,7 @@ static long mem_ctrl(BIO *b, int cmd, long num, void *ptr)
         ret = (long)bm->length;
         if (ptr != NULL) {
             pptr = (char **)ptr;
-            *pptr = (char *)&(bm->data[0]);
+            *pptr = (char *)(bm->data);
```
Comment (levitte) [inline crypto/bio/bss_mem.c:302]: Does this actually emit different code?

File: crypto/pem/pem_lib.c [hunk 2]
```diff
@@ -954,6 +954,11 @@ int PEM_read_bio_ex(BIO *bp, char **name_out, char **header,
     EVP_DecodeInit(ctx);
     BIO_get_mem_ptr(dataB, &buf_mem);
     len = buf_mem->length;
+
+    /* There was no data in the PEM file */
+    if (len == 0)
+        goto end;
+
```
Comment (levitte) [inline crypto/pem/pem_lib.c:961]: If you move down the EVP_ENCODE_CTX_new() call, you can have this length check even earlier...

File: crypto/bio/bss_mem.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/bio/bss_mem.c:302]: ubsan was complaining about the addition of 0 to a NULL pointer. So if `bm->data` is NULL then `&(bm->data[0])` is undefined!!

Comment (levitte) [reply crypto/bio/bss_mem.c:302]: Interesting. I thought most compilers optimizer away the "difference". Ok then

Comment (mattcaswell) [reply crypto/bio/bss_mem.c:302]: Well, yeah. I guess they do. But ubsan's job is to be picky about such things.

Comment (kroeckx) [reply crypto/bio/bss_mem.c:302]: The compiler is going to assume it has no undefined behaviour and compile both to the same thing. The reason to change the code is to make ubsan not give an error for the undefined behaviour.

Comment (levitte) [reply crypto/bio/bss_mem.c:302]: Oh of course. I was just a bit confused as to when that sanitizer operates...

File: crypto/pem/pem_lib.c (hunk 2 above)
Comment (mattcaswell) [reply crypto/pem/pem_lib.c:955]: Fixed

**Comment by mattcaswell** [comment]
Please take another look.

**Comment by levitte** [comment]
Did it fix what you expected?  The sanitizer build is still failing, but perhaps for other reasons?

**Comment by mattcaswell** [comment]
> Did it fix what you expected? The sanitizer build is still failing, but perhaps for other reasons?

This was targeting the run-checker ubsan issues which are completely different to the issues seen in the failing GitHub CI ubsan build. So I was never expecting it to solve the Github CI problems.

File: test/filterprov.c [hunk 3]
```diff
@@ -167,7 +167,7 @@ int filter_provider_set_filter(int operation, const char *filterstr)
     if (globs->num_dispatch >= MAX_FILTERS)
         goto err;
 
-    for (name = filterstrtmp; !last; name = sep + 1) {
+    for (name = filterstrtmp; !last; name = (last ? NULL : sep + 1)) {
```
Comment (levitte) [inline test/filterprov.c:170]: That expression had me scratch my head and wondering what it was all about.  Suggested alternative: `(sep == NULL ? NULL : sep + 1)`

**Review by levitte — CHANGES_REQUESTED** [review]

File: test/filterprov.c (hunk 3 above)
Comment (mattcaswell) [reply test/filterprov.c]: Done

**Comment by mattcaswell** [comment]
Feedback addressed.

**Review by levitte — APPROVED** [review]

File: crypto/bio/bss_mem.c (hunk 1 above)
Comment (paulidale) [inline crypto/bio/bss_mem.c:302]: Probably should drop the () around the pointer.

**Review by paulidale — APPROVED** [review]

File: crypto/bio/bss_mem.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/bio/bss_mem.c:302]: Hmm..I think it is clearer with?

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**